### PR TITLE
bpo-31334: Fix timeout in select.poll.poll()

### PR DESCRIPTION
**Pull request by volans-**
Always pass `-1`, or `INFTIM` where defined, to the `poll()` system call when
a negative timeout is passed to the `poll.poll([timeout])` method in the
`select` module. Various OSes throw an error with arbitrary negative
values.

https://bugs.python.org/issue31334

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by volans-** [comment]
For the record I've signed the CLA and my status has been updated on bugs.python.org.

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: Can't we do this in the following block?

```c
if (timeout_obj == NULL || timeout_obj == Py_None) {
    timeout = -1;
#ifdef INFTIM
    ms = INFTIM;
#else
    ms = -1;
#endif
    deadline = 0;
}
```

Then we can just change

```c
deadline = _PyTime_GetMonotonicClock() + timeout;
```

to

```c
if (ms >= 0) {
    deadline = _PyTime_GetMonotonicClock() + timeout;
}
```

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: See my comment about `INFTIM` below. I think we don't need to move this line out of this `if` branch.

File: Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:2]: Please add "Patch by Riccardo Coccioli.".

File: Lib/test/test_poll.py
Comment (berkerpeksag) [inline Lib/test/test_poll.py]: Style nitpick: I'd move `-1000` to the beginning of the tuple to make it ordered.

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: Another question: Can't we just set `timeout` to `INFTIM` or `-1` if `timeout_obj` is less than `-1` instead of messing with `ms`?

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Modules/selectmodule.c
Comment (volans-) [reply Modules/selectmodule.c]: First of all thanks for the review.

I'm not getting why this would be better. The current checks against `INT_MIN` and `INT_MAX` are done with `ms`, so I followed that pattern. Also `ms` is the actual parameter passed to the system call and the one causing the issue right now, it seemed more logic to me to change directly this one.

File: Modules/selectmodule.c
Comment (volans-) [reply Modules/selectmodule.c]: I don't see how what you're proposing can fix the issue. When a user passes a negative `timeout` parameter it enters the `else` block at line 542, so `ms` will be negative and the system call will still fail. The problem is not the `deadline` but the `ms` variable.

But your comment made me realise that I should add the `#ifdef` to the `if` block too, for coherence and in case an OS will ever have `INFTIM != -1`. To be DRY I'll move the `#ifdef` after the `else` block.

File: Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst
Comment (volans-) [reply Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:2]: Ok, thanks.

File: Lib/test/test_poll.py
Comment (volans-) [reply Lib/test/test_poll.py]: Ok.

**Comment by volans-** [comment]
@berkerpeksag did you had any chance to look at my replies and changes?

**Comment by volans-** [comment]
I didn't expect the Spanish Inquisition!

**Comment by volans-** [comment]
@berkerpeksag do you have any additional feedback on this change?

**Comment by berkerpeksag** [comment]
@volans- sorry for the delay. I need to take a closer look at this (and a BSD to test it locally) Please ping me again if I don't get to it by the next week.

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: We can set this in `deadline` initialization:

```c
_PyTime_t deadline = 0;
```

Comment (berkerpeksag) [reply Modules/selectmodule.c]: And the following comment is not needed anymore:

```c
/* initialize to prevent gcc warning */
```

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: Looking at this again, we can get rid of this if branch if we move out default values to variable declarations:

```c
_PyTime_t timeout = -1, ms = -1, deadline = 0;
```

Then we can get rid of the following comment too:

```c
/* Check values for timeout */
```

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: Style nit: Per PEP 7, we now use the following style in new C code:

```c
if (spam) {
    return 42;
}
````

instead of

```c
if (spam)
    return 42;
````

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c]: Style nit: Same as above.

Comment (berkerpeksag) [reply Modules/selectmodule.c]: Can we make this comment clearer? For example, it would be nice to mention this is the case if negative timeout is less than -1 on some OSes.

Also, I'd replace "many" with "some" or "some BSD OSes" since this is the first time this issue has been reported.

File: Lib/test/test_poll.py
Comment (berkerpeksag) [inline Lib/test/test_poll.py]: This is not directly related to this ticket, but I realized that `None` is not tested in `test_poll`. Could you also add `None` while we are here?

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

**Comment by volans-** [comment]
I have made the requested changes; please review again.

@berkerpeksag : Thank you very much for the review and the suggestions to simplify the code. My first patch was as less impactful as possible because I've found in other communities harder to get patches approved if they were also refactoring existing code. I'm happy to see a different approach here.

File: Modules/selectmodule.c
Comment (berkerpeksag) [inline Modules/selectmodule.c:559]: Nitpick: Can we change this to `ms < -1`? Since we set `ms = -1` in line 528, we shouldn't get any "invalid argument" exception if we don't pass a `timeout` value to `p.poll()`.

**Review by berkerpeksag — APPROVED** [review]
This looks pretty good to me, thank you! I just left a comment about the `ms < 0` check.

I will ask for another pair of eyes from another core developer then merge it.

File: Lib/test/test_poll.py
Comment (serhiy-storchaka) [inline Lib/test/test_poll.py]: I'm not sure that it is correct to insert infinite timeouts before finite timeouts in this test. Add them after `16000`. And maybe correct the multiplier in `(-1,)*10`.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Modules/selectmodule.c
Comment (volans-) [reply Modules/selectmodule.c:559]: @berkerpeksag what if a OS defines `INFTIM` different from `-1`? This ensures that we set `ms` to `INFTIM` in that case too. I can change it to `ms <= -1` if you think it's more readable.

On a side note, the documentation says only that the timeout parameter is in milliseconds, but doesn't say it must be an integer. I've noticed that passing a float works too and in the special case of passing a float in the range `(-1, 0)`, that is converted to `0` by `ms = _PyTime_AsMilliseconds(timeout, _PyTime_ROUND_CEILING);`. I've tested that this behaviour does not change with my patch, but is arguably the desired one.

Comment (serhiy-storchaka) [reply Modules/selectmodule.c:559]: I concur with @volans-. It is better to keep `ms < 0`.

As for rounding problem, I just have opened bpo-31786. I didn't mentioned it here because seems this change doesn't change this behavior.

Comment (berkerpeksag) [reply Modules/selectmodule.c:559]: @volans- fair enough, thanks for the explanation. I think `ms < 0` can stay as is, no need to change it to `ms <= -1`.

**Comment by volans-** [comment]
I have made the requested changes; please review again.

@serhiy-storchaka Thanks for the review. I hope that I correctly understood what you meant with:
> correct the multiplier in `(-1,)*10`.

I've also added a timeout smaller than the command output generation between the first one and the second one (`500`) and se the multiplier for the `-1` to `3`.
The result is that in one case the timeout ends before any output is generated (`500`), and in one case (the last `-1`) it checks that there is no output once the pipe is closed. In all the other cases one line of output is read.

File: Lib/test/test_poll.py
Comment (serhiy-storchaka) [inline Lib/test/test_poll.py]: I think that the total number of infinity timeouts should be not less than 10.  There are no guaranties that all finite timeouts are not expired before the subprocess will write the first line.

Comment (volans-) [reply Lib/test/test_poll.py]: @serhiy-storchaka Oh, sorry, now I got what you meant, updating the patch.

**Comment by berkerpeksag** [comment]
@serhiy-storchaka can you take a look at the last changes? I'd like to merge this if everything looks good to you.

**Comment by serhiy-storchaka** [comment]
An explicit test for infinity timeout will be added in #4003. It looks a better place for testing this PR. I suggest to wait until #4003 is merged.

File: Modules/selectmodule.c
Comment (vstinner) [inline Modules/selectmodule.c:564]: This normalization must be moved at the beginning of the loop below, since ms is recomputed if poll() is interrupted by a signal (EINTR).

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Modules/selectmodule.c
Comment (volans-) [reply Modules/selectmodule.c:564]: @haypo where do you mean exactly? `ms` is recalculated only if `timeout` is non-negative and `ms` cannot be negative with a non-negative `timeout`.

Comment (serhiy-storchaka) [reply Modules/selectmodule.c:564]: I agree with @volans-. No changes are needed here.

File: Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:2]: nitpick: "non-negative" is better known as "positive" :-)

File: Modules/selectmodule.c
Comment (vstinner) [reply Modules/selectmodule.c:564]: Right.

**Review by vstinner — APPROVED** [review]

**Review by vstinner — APPROVED** [review]
LGTM, except of a minor nitpick.

File: Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst
Comment (volans-) [reply Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:2]: @haypo  "positive" doesn't include zero... :-)

Comment (vstinner) [reply Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:2]: "strictly positive" doesn't include zero, "positive" does. Well, I don't really care. The current sentence is correct.

Comment (serhiy-storchaka) [reply Misc/NEWS.d/next/Library/2017-09-04-00-22-31.bpo-31334.9WYRfi.rst:2]: "non-negative" means "positive or zero".

File: Lib/test/test_poll.py
Comment (serhiy-storchaka) [inline Lib/test/test_poll.py]: #4003 is merged. Please revert this change and add the timeout value -1000 in the new test method.

Comment (volans-) [reply Lib/test/test_poll.py]: @serhiy-storchaka sure, will do.
What is the preferred method to update this branch with the new upstream code, `git merge`? I'm asking because this is my first patch and the developer guide mention to not use _anything that would require a force push to GitHub_, so I guess a rebase is excluded.

Comment (serhiy-storchaka) [reply Lib/test/test_poll.py]: Yes, just use `git merge`.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
Wow, 3 approvals from 3 different core developers. Well done @volans- ! Thanks for this nice bug fix ;-)

**Comment by vstinner** [comment]
I removed the "needs backport" labels to prevent the bot to try again to create backport fixes.

**Comment by volans-** [comment]
@haypo thank you all for the review, comments and for approving/merging it.

Should I create manual PRs for the backports given that the bot seems to have some issues today?

**Comment by serhiy-storchaka** [comment]
First #4003 should be backported.

**Comment by volans-** [comment]
@serhiy-storchaka I've ported this to the 2.7 branch with GH-4034, it's a smaller patch given the code in the 2.7 branch.

**Comment by serhiy-storchaka** [comment]
Thanks @volans-!

**Pull request merged**
